README.md
[![Build Status](https://travis-ci.com/hengshiyu/cvcrand.svg?token=ZtDwjW3Z5FqDgnzwZCn6&branch=master)](https://travis-ci.com/hengshiyu/cvcrand)

# cvcrand R-package for Covariate-constrained Randomization and Clustered Permutation Test for Cluster Randomized Trials
Hengshi Yu, Fan Li, John A. Gallis and Elizabeth L. Turner

**Maintainer**: Hengshi Yu <hengshi@umich.edu>

## Introduction
cvcrand is an R package for the design and analysis of cluster randomized trials (CRTs). 

A cluster is the unit of randomization for a cluster randomized trial. Thus, when the number of clusters is small, there might be some baseline imbalance from the randomization between the arms. Constrained randomization constrained the randomization space. Given the baseline values of some cluster-level covariates, users can perform a constrained randomization on the clusters into two arms, with an optional input of user-defined weights on the covariates. 

At the end of the study, the individual outcome is collected. The `cvcrand` package then performs clustered permutation test on either continuous outcome or binary outcome adjusted for some individual-level covariates, producing p-value of the intervention effect.

## Functions and references
The cvcrand package constains two main functions. In the design of CRTs with two arms, users can use the `cvcrand()` function to perform constrained randomization. And for the analysis part, user will use the `cptest()` function for clustered permutation test. 

1. cvcrand function: constrained randomization for two-arm cluster randomized trials
    * Raab, G.M. and Butcher, I., 2001. Balance in cluster randomized trials. Statistics in medicine, 20(3), pp.351-365.
    * Li, F., Lokhnygina, Y., Murray, D.M., Heagerty, P.J. and DeLong, E.R., 2016. An evaluation of constrained randomization for the design and analysis of group‐randomized trials. Statistics in medicine, 35(10), pp.1565-1579.
    * Li, F., Turner, E. L., Heagerty, P. J., Murray, D. M., Vollmer, W. M., & DeLong, E. R. (2017). An evaluation of constrained randomization for the design and analysis of group‐randomized trials with binary outcomes. Statistics in medicine, 36(24), 3791-3806.
    * Dickinson, L. M., Beaty, B., Fox, C., Pace, W., Dickinson, W. P., Emsermann, C., & Kempe, A. (2015). Pragmatic cluster randomized trials using covariate constrained randomization: A method for practice-based research networks (PBRNs). The Journal of the American Board of Family Medicine, 28(5), 663-672.
    


2. cptest function: clustered permutation test for two-arm cluster randomized trial
    * Gail, M.H., Mark, S.D., Carroll, R.J., Green, S.B. and Pee, D., 1996. On design considerations and randomization‐based inference for community intervention trials. Statistics in medicine, 15(11), pp.1069-1092.
    * Li, F., Lokhnygina, Y., Murray, D.M., Heagerty, P.J. and DeLong, E.R., 2016. An evaluation of constrained randomization for the design and analysis of group‐randomized trials. Statistics in medicine, 35(10), pp.1565-1579.
    * Li, F., Turner, E. L., Heagerty, P. J., Murray, D. M., Vollmer, W. M., & DeLong, E. R. (2017). An evaluation of constrained randomization for the design and analysis of group‐randomized trials with binary outcomes. Statistics in medicine, 36(24), 3791-3806.
    * Dickinson, L. M., Beaty, B., Fox, C., Pace, W., Dickinson, W. P., Emsermann, C., & Kempe, A. (2015). Pragmatic cluster randomized trials using covariate constrained randomization: A method for practice-based research networks (PBRNs). The Journal of the American Board of Family Medicine, 28(5), 663-672.
    * Eldridge, S. M., Ukoumunne, O. C., & Carlin, J. B. (2009). The Intra‐Cluster Correlation Coefficient in Cluster Randomized Trials: A Review of Definitions. International Statistical Review, 77(3), 378-394.
    * Hannan, P. J., Murray, D. M., Jacobs Jr, D. R., & McGovern, P. G. (1994). Parameters to aid in the design and analysis of community trials: intraclass correlations from the Minnesota Heart Health Program. Epidemiology, 88-95.


### `cvcrand()` example: covariate-constrained randomization

The balance score for constrained randomization in the program is developed from Raab and Butcher (2001).  

A study presented by Dickinson et al (2015) is about two approaches (interventions) for increasing the "up-to-date" immunization rate in 19- to 35-month-old children. They planned to randomize 16 counties in Colorado 1:1 to either a population-based approach or a practice-based approach. There are several county-level variables. The program will randomize on a subset of these variables. The continuous variable of average income is categorized to illustrate the use of the `cvcrand()` on multi-category variables. And the percentage in Colorado Immunization Information System (CIIS) variable is truncated at 100%.

For the constrained randomization, we used the `cvcrand()` function to randomize 8 out of the 16 counties into the practice-based. For the definition of the whole randomization space, if the total number of all possible schemes is smaller than `50,000`, we enumerate all the schemes as the whole randomization space. Otherwise, we simulate `50,000` schemes and choose the unique shemes among them as the whole randomization space. We calculate the balance scores of `"l2"` metric on three continuous covariates as well as two categorical covariates of location and income category. Location has `"Rural"` and `"Urban"`. The level of `"Rural"` was then dropped in `cvcrand()`. As income category has three levels of `"low"`, `"med"`, and `"high"`,  the level of `"high"` was dropped to create dummy variables according to the alphanumerical order as well. Then we constrained the randomization space to the schemes with `"l2"` balance scores less than the `0.1` quantile of that in the whole randomization space. Finally, a randomization scheme is sampled from the constrained space.  

We saved the constrained randomization space in a CSV file in `"dickinson_constrained.csv"`, the first column of which is an indicator variable of the finally selected scheme (`1`) or not (`0`). We also saved the balance scores of the whole randomization space in a CSV file in `"dickinson_bscores.csv"`, and output a histogram displaying the distribution of all balance scores with a red line indicating our selected cutoff (the `0.1` quantile).


```r
 Design_result <- cvcrand(clustername = Dickinson_design$county,
                  balancemetric = "l2",
                  x = data.frame(Dickinson_design[ , c("location", "inciis",
                      "uptodateonimmunizations", "hispanic", "incomecat")]),
                  ntotal_cluster = 16,
                  ntrt_cluster = 8,
                  categorical = c("location", "incomecat"),
                  savedata = "dickinson_constrained.csv",
                  savebscores = "dickinson_bscores.csv",
                  cutoff = 0.1,
                  seed = 12345)
```


### `cvcrand()` example: stratified constrained randomization

User-defined weights can be used to induce stratification on one or more categorical variables. In the study presented by Dickinson et al (2015), there are 8 `"Urban"` and 8 `"Rural"` counties. A user-defined weight of `1,000` is added to the covariate of `location`, while these weights for other covariates are all `1`. Intuitively, a large weight assigned to a covariate sharply penalizes any imbalance of that covariates, therefore including schemes that are optimally balanced with respect to that covariate in the constrained randomization space. In practice, the resulting constrained space approximates the stratified randomization space on that covariate. In our illustrative data example, since half of the counties are located in rural areas, perfect balance is achieved by considering constrained randomization with the large weight for `location` variable. Alternatively, the option of `stratify` is able to perform the equivalent stratification on the stratifying variables specified.


```r
# Stratification on location

Design_stratified_result1 <- cvcrand(clustername = Dickinson_design$county,
                                     balancemetric = "l2",
                                     x = data.frame(Dickinson_design[ , c("location", "inciis",
                                     "uptodateonimmunizations", 
                                     "hispanic", "incomecat")]),
                                     ntotal_cluster = 16,
                                     ntrt_cluster = 8,
                                     categorical = c("location", "incomecat"),
                                     weights = c(1000, 1, 1, 1, 1),
                                     cutoff = 0.1,
                                     seed = 12345) 
                                  
                                  
# An alternative and equivalent way to stratify on location

Design_stratified_result2 <- cvcrand(clustername = Dickinson_design$county,
                                     balancemetric = "l2",
                                     x = data.frame(Dickinson_design[ , c("location", "inciis",
                                     "uptodateonimmunizations", 
                                     "hispanic", "incomecat")]),
                                     ntotal_cluster = 16,
                                     ntrt_cluster = 8,
                                     categorical = c("location", "incomecat"),
                                     stratify = "location",
                                     cutoff = 0.1,
                                     seed = 12345)

```


### `cptest()` example: Clustered Permutation Test

At the end of cluster randomized trials, individual outcomes are collected. Permutation test based on Gail et al (1996) and Li et al (2016) is then applied to the continuous or binary outcome with some individual-level covariates. 

Suppose that the researchers were able to assess 300 children in each cluster in a study presented by Dickinson et al (2015), and the cluster randomized trial is processed with the selected randomization scheme from the example above of the `cvcrand()` function. We expanded the values of the cluster-level covariates on the covariates' values of the individuals, according to which cluster they belong to. The correlated individual outcome of up-to-date on immunizations (`1`) or not (`0`) is then simulated using a generalized linear mixed model (GLMM) with a logistic link to induce correlation by including a random effect at the county level. The intracluster correlation (ICC) was set to be 0.01, using the latent response definition provided in Eldridge et al (2009). This is a reasonable value for population health studies Hannan et al (1994). We simulated one data set, with the outcome data dependent on the county-level covariates used in the constrained randomization design and a positive treatment effect so that the practice-based intervention increases up-to-date immunization rates more than the community-based intervention. For each individual child, the outcome is equal to `1` if he or she is up-to-date on immunizations and `0` otherwise. 

We used the `cptest()` function to process the clustered permutation test on the binary outcome of the status of up-to-date on immunizations. We input the file about the constrained space with the first column indicating the final scheme. The permutation test is on the continuous covariates of `"inciis"`, `"uptodateonimmunizations"`, `"hispanic"`, as well as categorical variables of `"location"` and `"incomecat"`. Location has `"Rural"` and `"Urban"`. The level of `"Rural"` was then dropped in `cptest()`. As income category has three levels of `"low"`, `"med"`, and `"high"`,  the level of `"high"` was dropped to create dummy variables according to the alphanumerical order as well.

```r
 Analysis_result <- cptest(outcome = Dickinson_outcome$outcome,
                           clustername = Dickinson_outcome$county,
                           z = data.frame(Dickinson_outcome[ , c("location", "inciis",
                               "uptodateonimmunizations", "hispanic", "incomecat")]), 
                           cspacedatname = "dickinson_constrained.csv",                                 
                           outcometype = "binary",                                                      
                           categorical = c("location","incomecat"))
```

## Installation

The `cvcrand` R package is available on CRAN.


NEWS.md
# cvcrand 0.1.1

## Major changes

Fix the problem of broken @docType package documentation

R/cvcrand_package.R
#'cvcrand: a package for efficient design and analysis of cluster randomized trials
#'@section cvcrand functions:
#'The cvrall function performs constrained randomization on cluster randomized trials (CRTs)
#'The cvrcov function performs covariate-constrained randomization on cluster randomized trials (CRTs)
#'The cptest function performs permutation test on the outcome of cluster randomized trials (CRTs)
#'@docType package
#'@name cvcrand
#'@import tableone
#'@importFrom graphics abline box hist mtext par title
#'@importFrom stats as.formula glm lm model.matrix quantile sd residuals
#'@importFrom utils combn read.csv write.csv capture.output
#'@description cvcrand: A Package for Covariate-constrained Randomization and the Clustered Permutation Test for
#'Cluster Randomized Trials
#' @keywords internal
"_PACKAGE"



R/Dickinson_design.R
#' @name Dickinson_design
#' @title Raw county-level variables for study 1 in Dickinson et al (2015)
#' @description
#'   Two approaches (interventions) were compared for increasing the "up-to-date" immunization rate
#'  in 19- to 35-month-old children. 16 counties in Colorado 1:1 were randomized 1:1 to either a
#'  population-based approach or a practice-based approach. Ahead of randomization, several
#'  county-level variables were collected, and a subset of them are used for covariate constrained
#'  randomization. The continuous variable of average income is categorized into tertiles to illustrate the
#'  use of cvcrand on multi-category variables. And the percentage in CIIS variable is truncated at 100%.
#' @docType data
#' @format A data frame with 16 rows and 11 variables:
#' \describe{
#'   \item{county}{the identification for the county}
#'   \item{location}{urban or rural}
#'   \item{inciis}{percentage of children aged 19-35 months in the Colorado Immunization Information System (CIIS)}
#'   \item{numberofchildrenages1935months}{number of children aged 19-35 months}
#'   \item{uptodateonimmunizations}{percentage of children already up-to-date on their immunization}
#'   \item{africanamerican}{percentage of population that is African American}
#'   \item{hispanic}{percentage of population that is Hispanic}
#'   \item{income}{average income}
#'   \item{incomecat}{average income categorized into tertiles}
#'   \item{pediatricpracticetofamilymedicin}{pediatric practice-to-family medicine practice ratio}
#'   \item{communityhealthcenters}{number of community health centers}
#' }
#' @source \url{https://www.jabfm.org/content/28/5/663/tab-figures-data}
#' @references
#'   Dickinson, L. M., B. Beaty, C. Fox, W. Pace, W. P. Dickinson, C. Emsermann,
#'   and A. Kempe (2015): Pragmatic cluster randomized trials using covariate
#'   constrained randomization: A method for practice-based research networks (PBRNs).
#'   The Journal of the American Board of Family Medicine 28(5): 663-672
NULL


R/Dickinson_outcome.R
#' @name Dickinson_outcome
#' @title Simulated individual-level binary outcome and baseline variables for study 1 in Dickinson et al (2015)
#' @description
#'   At the end of the study, the researchers will have ascertained the outcome in the
#'   16 clusters. Suppose that the researchers were able to assess 300 children in each cluster.
#'   We simulated correlated outcome data at the individual level using a generalized linear
#'   mixed model (GLMM) to induce correlation by including a random effect. The intracluster
#'   correlation (ICC) was set to be 0.01, using the latent response definition provided in
#'   Eldrige et al. (2009). This is a reasonable value of the ICC for population health studies
#'   (Hannan et al. 1994). We simulated one data set, with the outcome data dependent on the
#'   county-level covariates used in the constrained randomization design and a positive treatment
#'   effect so that the practice-based intervention increases up-to-date immunization rates more
#'   than the community-based intervention. For each individual child, the outcome is equal to 1
#'   if he or she is up-to-date on immunizations and 0 otherwise.
#'
#'   Note that we still categorize the continuous variable of average income to illustrate
#'   the use of cvcrand on multi-category variables, and we truncated the percentage
#'   in CIIS variable at 100%.
#' @docType data
#' @format A data frame with 4800 rows and 7 variables:
#' \describe{
#'   \item{county}{the identification for the county}
#'   \item{location}{urban or rural}
#'   \item{inciis}{percentage of children ages 19-35 months in the Colorado Immunization Information System (CIIS)}
#'   \item{uptodateonimmunizations}{percentage of children already up-to-date on their immunization}
#'   \item{hispanic}{percentage of population that is Hispanic}
#'   \item{incomecat}{average income categorized into tertiles}
#'   \item{outcome}{the status of being up-to-date on immunizations}
#' }
#' @references
#'   Dickinson, L. M., B. Beaty, C. Fox, W. Pace, W. P. Dickinson, C. Emsermann,
#'   and A. Kempe (2015): Pragmatic cluster randomized trials using covariate
#'   constrained randomization: A method for practice-based research networks (PBRNs).
#'   The Journal of the American Board of Family Medicine 28(5): 663-672
#'
#'   Eldridge, S. M., Ukoumunne, O. C., & Carlin, J. B. (2009). The Intra Cluster
#'   Correlation Coefficient in Cluster Randomized Trials: A Review of Definitions.
#'   International Statistical Review, 77(3), 378-394.
#'
#'   Hannan, P. J., Murray, D. M., Jacobs Jr, D. R., & McGovern, P. G. (1994).
#'   Parameters to aid in the design and analysis of community trials: intraclass
#'   correlations from the Minnesota Heart Health Program. Epidemiology, 88-95. ISO 690
NULL


R/cptest.R
#' Clustered permutation test for cluster randomized trials
#' @param outcome a vector specifying the individual-level outcome.
#' @param clustername a vector specifying the identification variable of the cluster.
#' @param z  a data frame of covariates to be adjusted for in the permutation analysis.
#' @param cspacedatname  gives the path of the csv dataset containing the saved randomization space. This dataset contains the permutation matrix, as well as an indicator variable in the first column indicating which row of the permutation matrix was selected as the final scheme to be implemented in practice.
#' @param outcometype the type of regression model that should be run. Options are \code{"continuous"} for linear regression and \code{"binary"} for logistic regression.
#' @param categorical a vector specifying categorical (including binary) variables. This can be names of the columns or number indexes of columns, but cannot be both. Suppose there are \code{p} categories for a categorical variable, \code{cptest} function creates \code{(p-1)} dummy variables and drops the reference level if the variable is specified as a factor. Otherwise, the first level in the alphanumerical order will be dropped. If the user wants to specify a different level to drop for a \code{p}-level categorical variable, the user can create \code{p-1} dummy variables and these can instead be supplied as covariates to the \code{cptest} function. Then, the user needs to specify the dummy variables created themselves to be \code{categorical} when running \code{cptest}. In addition, the user could also set the variable as a factor with the specific reference level. The user must ensure that the same level of the categorical variable is excluded as was excluded when running \code{cvrall}, by coding the variables the same way as in the design phase. This is the only optional argument of the \code{cptest} function. All others are required.
#' @keywords cluster-randomized-trails clustered-permutation-test
#' @author Hengshi Yu <hengshi@umich.edu>, Fan Li <fan.f.li@yale.edu>, John A. Gallis <john.gallis@duke.edu>, Elizabeth L. Turner <liz.turner@duke.edu>
#' @description cptest performs a clustered permutation test on the individual-level outcome data for cluster
#' randomized trials (CRTs). The type of the outcome can be specified by the user to be \code{"continuous"} or
#' \code{"binary"}.
#'
#' Linear regression (for outcome type \code{"continuous"}) or logistic regression (for outcome type \code{"binary"}) is applied to the outcome regressed on covariates specified. Cluster residual means are computed. Within the constrained space,
#' the contrast statistic between the treatment and control arms is created from the randomization schemes and the cluster residual means. The permutation test is then conducted by comparing the contrast statistic for the scheme actually utilized to all other schemes in the constrained space.
#' @references
#' Gail, M.H., Mark, S.D., Carroll, R.J., Green, S.B. and Pee, D., 1996. On design considerations and randomization based inference for community intervention trials. Statistics in medicine, 15(11), pp.1069-1092.
#'
#' Li, F., Lokhnygina, Y., Murray, D.M., Heagerty, P.J. and DeLong, E.R., 2016. An evaluation of constrained randomization for the design and analysis of group randomized trials. Statistics in medicine, 35(10), pp.1565-1579.
#'
#' Li, F., Turner, E. L., Heagerty, P. J., Murray, D. M., Vollmer, W. M., & DeLong, E. R. (2017). An evaluation of constrained randomization for the design and analysis of group randomized trials with binary outcomes. Statistics in medicine, 36(24), 3791-3806.
#'
#' Gallis, J. A., Li, F., Yu, H., Turner, E. L. (In Press).  cvcrand and cptest: Efficient design and analysis of cluster randomized trials.  Stata Journal.
#'
#' Gallis, J. A., Li, Fl. Yu, H., Turner, E. L. (2017).  cvcrand and cptest: Efficient design and analysis of cluster randomized trials.  Stata Conference.  https://www.stata.com/meeting/baltimore17/slides/Baltimore17_Gallis.pdf.
#'
#' Dickinson, L. M., Beaty, B., Fox, C., Pace, W., Dickinson, W. P., Emsermann, C., & Kempe, A. (2015). Pragmatic cluster randomized trials using covariate constrained randomization: A method for practice-based research networks (PBRNs). The Journal of the American Board of Family Medicine, 28(5), 663-672.
#' @export
#' @examples
#'\dontrun{
#' Analysis_result <- cptest(outcome = Dickinson_outcome$outcome,
#'                           clustername = Dickinson_outcome$county,
#'                           z = data.frame(Dickinson_outcome[ , c("location", "inciis",
#'                               "uptodateonimmunizations", "hispanic", "incomecat")]),
#'                           cspacedatname = "dickinson_constrained.csv",
#'                           outcometype = "binary",
#'                           categorical = c("location","incomecat"))
#'}
#'
#' @return \code{FinalScheme} the final scheme in the permutation matrix
#' @return \code{pvalue} the p-value of the intervention effect from the clustered permutation test
#' @return \code{pvalue_statement} the statement about the p-value of the intervention effect from the clustered permutation test






cptest <- function(outcome, clustername, z = NULL, cspacedatname, outcometype, categorical = NULL){
  x <- z
  if (!is.null(z)) {
       if (is.null(categorical)) { # if there are no categorical variables specified

   if (sum(apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1) >= 1) {
     ## check the variables with only one unique value, i.e. no variation

     warning(cat("Warning: each of columns", c(1 : dim(x)[2])[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1], "has only less than or equal to one unique value and no variation."))
   }




   if (!setequal(c(1 : dim(x)[2])[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 4], categorical)) {
     ## check the categorical variables have been correctly specified

     warning(cat("Warning: each of columns", c(1 : dim(x)[2])[apply(x, 2, function(x) length(unique(x))) <= 4], "has less than or equal to 4 unique observations. Please check all the categorical variables have been correctly specified."))


   }


    } else {  # If there are categorical variables specified

    p <- length(categorical)

     if (is.character(categorical)) {  ## columns names based categorical variables

      if (sum(apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1) >= 1) {
         ## check the variables with only one unique value, i.e. no variation

         warning(cat("Warning: each of", colnames(x)[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1], "has only less than or equal to one unique value and no variation."))
       }


      if (!setequal(colnames(x)[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 4], categorical)) {
        ## check the categorical variables have been correctly specified


        warning(cat("Warning: each of", setdiff(colnames(x)[apply(x, 2, function(x) length(unique(x))) <= 4], categorical), "has less than or equal to 4 unique observations. Please check all the categorical variables have been correctly specified."))
      }



       for(i in 1:p){

             x <- data.frame(x, model.matrix(~factor(x[ , categorical[i]]))[ , -1])
                           # add the dummy variables into the x matrix or data frame
                     }


        x <- x[ , -which(colnames(x) %in% categorical)]
                           # remove the initial categorical variables from x

          } else if (is.numeric(categorical)) { ## columns indexes based categorical variables

            if(sum(apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1) >= 1){
             ## check the variables with only one unique value, i.e. no variation

           warning(cat("Warning: each of columns", c(1:dim(x)[2])[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 1], "has only less than or equal to one unique value and no variation."))

          }


         if (!setequal(c(1:dim(x)[2])[apply(x, 2, function(x) length(unique(x[!is.na(x)]))) <= 4], categorical)) {
                           ## check the categorical variables have been correctly specified

       warning(cat("Warning: each of columns",setdiff(c(1:dim(x)[2])[apply(x, 2, function(x) length(unique(x))) <= 4], categorical), "has less than or equal to 4 unique observations, Please check all the categorical variables have been correctly specified."))


     }



         for (i in 1:p) {

               x <- data.frame(x, model.matrix(~factor(x[ , categorical[i]]))[ ,-1])

                        } # adding the dummy variables into the x matrix or data frame

          x <- x[ , -categorical] # remove the initial categorical variables
   }

  }
  
  }
 


  pmt <- read.csv(cspacedatname, header = TRUE)
  # read in the data

  rw <- which(pmt[ , 1] == 1)
  # find the chosen scheme


  tc <- pmt[rw, -1]
  # the chosen scheme


  dpmt <- pmt[ , -1]
  # all the schemes without the chosen indicator variable


  for (i in 1:dim(dpmt)[1]) {

  	        for (j in 1:dim(dpmt)[2]) {

                        if (dpmt[i, j]==0) {

           	                            dpmt[i, j] <- -1

                        }

  	        }

  }
    # transform 0 in the constrained space into -1



  if (!all(apply(dpmt, 1, sum)==0)) {

    warning("Warning: test might be anti-conservative if number of intervention clusters does not equal number of control clusters.")
    ## If there is any scheme with unequal cluster assignment, the cptest function has inflated test size.

  }



  if (!is.null(z)) {
    fm <- as.formula(paste0("outcome~", paste(names(x), collapse = "+")))
  } else {
    fm <- as.formula("outcome ~ 1")
  }
   
   # the formula of the model

   x$outcome <- outcome
   # merge the outcome and the covariates into one data frame

   if (outcometype == "continuous") {

       ADJMeans <- tapply(lm(formula = fm, data = x)$res, clustername, mean)
       # for continuous outcome, we use linear regression

       Diffs <- as.matrix(dpmt) %*% ADJMeans
       # the permutation statistic

       pvalue <- mean(abs(Diffs) >= abs(Diffs[rw, ]))

   } else if (outcometype == "binary") {

       ADJMeans <- tapply(residuals(glm(formula = fm, family = "binomial", data = x), type = "response"), clustername, mean)
       # for binary outcome, we use logistic regression

       Diffs <- as.matrix(dpmt) %*% ADJMeans
       # the permutation statistic

       pvalue <- mean(abs(Diffs) >= abs(Diffs[rw, ]))

   } else {

   	stop("Error: Please specify the correct outcometype for continuous or binary")
    # for other type outcome, just cannot run the cptest function

   }

 FinalScheme <- data.frame(Cluster_ID = names(ADJMeans), Intervention = as.vector(as.matrix(tc)))
 # the final chosen scheme in the constrained space

 pvalue_message <- paste("Clustered permutation test p-value =", round(pvalue, 4), collapse=' ')
 # the p-value from the permutation test

  return(list(FinalScheme = FinalScheme,
              pvalue = round(pvalue, 4),
  	          pvalue_statement = pvalue_message))

  }




